From: Christian Clason <c.clason@uni-graz.at>
Date: Tue, 31 Jan 2023 17:38:47 +0100
X-Dgit-Generated: 0.7.2-7 ce58d4397e8cc3e7e3ba284a70068f51e29fe8da
Subject: fix(test): adapt test to release-0.8

Signed-off-by: James McCoy <jamessan@debian.org>

---

diff --git a/test/functional/treesitter/language_spec.lua b/test/functional/treesitter/language_spec.lua
index c825a33d5..e183794eb 100644
--- a/test/functional/treesitter/language_spec.lua
+++ b/test/functional/treesitter/language_spec.lua
@@ -29,7 +29,7 @@ describe('treesitter API', function()
   end)
 
   it('shows error for invalid language name', function()
-    eq(".../language.lua:0: '/foo/' is not a valid language name",
+    eq("Error executing lua: .../language.lua:0: '/foo/' is not a valid language name",
       pcall_err(exec_lua, 'vim.treesitter.require_language("/foo/", nil, false)'))
   end)
 
